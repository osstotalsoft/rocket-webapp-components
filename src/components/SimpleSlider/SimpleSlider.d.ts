import { FormControlProps } from '@material-ui/core/FormControl'
import { FormHelperTextProps } from '@material-ui/core/FormHelperText'

export interface Props {

    /**
     * The value of the label
     */
    label?: string;

    /**
     * The value of the slider. For ranged sliders, provide an array with two values.
     */
    value?: number | number[];

    /**
     * Callback function that is fired when the slider's value changed.
     * event: The event source of the callback.
     * value: The new value.
     */
    onChange?: (event: object, value: number | number[]) => void;

    /**
     * If true, the component will take up the full width of its container.
     */
    fullWidth?: boolean;

    /**
     * The minimum allowed value of the slider. Should not be equal to max.
     */
    min?: number;

    /**
     * The maximum allowed value of the slider. Should not be equal to min.
     */
    max?: number;

    /**
     * The granularity with which the slider can step through values. (A "discrete" slider.)
     * The min prop serves as the origin for the valid values. We recommend (max - min) to be evenly divisible by the step.
     * When step is null, the thumb can only be slid onto marks provided with the marks prop.
     */
    step?: number;

    /**
     * Marks indicate predetermined values to which the user can move the slider. If true the marks will be spaced according 
     * the value of the step prop. If an array, it should contain objects with value and an optional label keys.
     */
    marks?: boolean | array;

    /**
     * If true, the error data is provided.
     */
    error?: boolean;

    /**
     * The helper text content.
     */
    helperText?: string;

    /**
     * If true, the label is displayed as required and the input element` will be required.
     */
    required?: boolean;

    /**
     * number of decimals to be displayed
     */
    decimalScale?: number;
  
    /**
    * Add thousand separators on number; single character string or boolean true (true is default to ,)
    */
    thousandSeparator?: boolean | string;

    /**
     * If true, the slider will be disabled
     */
    disabled?: boolean;

    /**
     * If true, the slider limits will be displayed
     */
    showSliderLimits?: boolean;

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color?: 'primary' | 'secondary';

    /**
     * The slider orientation.
     */
    orientation?: 'horizontal' | 'vertical';

    /**
     * Demos:
     * 
     * - https://bit.dev/totalsoft_oss/react-mui/simple-slider
     * 
     *  Take a look over the code for further information:https://github.com/osstotalsoft/rocket-webapp-components/blob/master/components/SimpleSlider/SimpleSlider.js
     * 
     * Material-UI component used:
     * 
     * - Slider: https://material-ui.com/api/slider/
     * 
     */

}

export type SimpleSliderProps = Props | FormControlProps | FormHelperTextProps

export default function SimpleSlider(props: SimpleSliderProps): JSX.Element;
