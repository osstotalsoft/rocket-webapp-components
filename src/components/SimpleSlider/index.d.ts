export { default } from './SimpleSlider';
export * from './SimpleSlider';