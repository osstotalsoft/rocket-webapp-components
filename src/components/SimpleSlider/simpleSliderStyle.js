const simpleSliderStyles = {
    sliderRoot: {
        top: "-1px",
        margin: 0,
        padding: 0
    },
    sliderTrackError: {
        backgroundColor: "red"
    },
    sliderThumb: {
        zIndex: 0,
    },
    disabledText: {
        opacity: 0.6
    },
    //fix for mobile horizontal scroll
    formControl: {
        overflow: "visible"
    },
    sliderLimits: {
        display: "flex",
        justifyContent: "space-between"
    }
};

export default simpleSliderStyles