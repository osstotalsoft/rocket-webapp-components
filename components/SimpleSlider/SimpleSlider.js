import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import { FormControl, FormHelperText } from '@material-ui/core';
import CustomTextField from "../CustomTextField";
import Typography from '../Typography';
import Slider from '@material-ui/core/Slider';
import debounce from 'lodash.debounce';
import simpleSliderStyle from "./simpleSliderStyle";

const useStyles = makeStyles(simpleSliderStyle);

const SimpleSlider = ({ label, fullWidth, value, min, max, step, onChange, error,
    helperText, required, decimalScale, thousandSeparator, disabled, showSliderLimits, ...other }) => {
    const classes = useStyles();

    const onTextChanged = useCallback(event => {
        if (event.target && event.target.value) {
            if (event.target.value <= min) {
                onChange(null, min);
            }
            else if (event.target.value >= max) {
                onChange(null, max);
            }
            else {
                onChange(null, event.target.value);
            }
        }
        else {
            onChange(null, min);
        }
    }, [onChange, min, max])

    const OnTextChangedDebounced = debounce(onTextChanged, 1000)

    return (
        <FormControl fullWidth={fullWidth} className={classes.formControl} error={error} required={required}>
            <CustomTextField
                label={label}
                fullWidth={fullWidth}
                required={required}
                value={value}
                onChange={OnTextChangedDebounced}
                isNumeric={true}
                inputProps={{ decimalScale: decimalScale, thousandSeparator: thousandSeparator }}
                disabled={disabled}
            />

            <Slider classes={{ root: classes.sliderRoot, track: error ? classes.sliderTrackError : null, thumb: classes.sliderThumb }}
                value={value}
                min={min}
                max={max}
                step={step}
                aria-labelledby="continuous-slider"
                onChange={onChange}
                disabled={disabled}
                {...other}
            />

            {showSliderLimits && (
                <div className={classes.sliderLimits}>
                    <Typography variant="overline" className={classes.disabledText}>{min ? min.toFixed(decimalScale) : min}</Typography>
                    <Typography variant="overline" className={classes.disabledText} style={{ textAlign: 'end' }}>{max ? max.toFixed(decimalScale) : max}</Typography>
                </div>
            )}
            <FormHelperText error={error}>{helperText}</FormHelperText>
        </FormControl>
    )
}

SimpleSlider.propTypes = {
    label: PropTypes.string,
    value: PropTypes.any,
    onChange: PropTypes.func,
    fullWidth: PropTypes.bool,
    min: PropTypes.number,
    max: PropTypes.number,
    step: PropTypes.number,
    error: PropTypes.bool,
    helperText: PropTypes.string,
    required: PropTypes.bool,
    decimalScale: PropTypes.number,
    thousandSeparator: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    disabled: PropTypes.bool,
    showSliderLimits: PropTypes.bool
};

export default SimpleSlider;
