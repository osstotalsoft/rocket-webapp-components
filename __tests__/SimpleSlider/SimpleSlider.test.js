import React from "react";
import renderer from "react-test-renderer";
import SimpleSlider from "../../components/SimpleSlider/SimpleSlider";

jest.mock('react-dom', () => ({
  findDOMNode: (instance) => {
    return { ownerDocument: instance };
  },
}));

describe("SimpleSlider", () => {
  it("Snapshot", () => {
    let eventListenerFn = jest.fn();
    const component = renderer.create(
      <SimpleSlider
        fullWidth
        required
        value={null}
        min={0}
        max={100}
        step={1}
        label={"Simple slider"}
        disabled={false}
        onChange={() => { }}
        decimalScale={2}
        showSliderLimits={true}
      />, {
      createNodeMock: (element) => {
        if (element.type === 'span') {
          return {
            addEventListener: eventListenerFn,
          };
        }
      }
    }
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
