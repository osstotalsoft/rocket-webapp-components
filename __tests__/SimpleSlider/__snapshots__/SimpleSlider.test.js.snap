// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleSlider Snapshot 1`] = `
<div
  className="MuiFormControl-root makeStyles-formControl-5 MuiFormControl-fullWidth"
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root makeStyles-label-7 MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
      data-shrink={false}
    >
      Simple slider
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
        â€‰
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input makeStyles-input-8 "
        decimalScale={2}
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        type="text"
        value={null}
      />
    </div>
  </div>
  <span
    className="MuiSlider-root makeStyles-sliderRoot-1 MuiSlider-colorPrimary"
    onMouseDown={[Function]}
  >
    <span
      className="MuiSlider-rail"
    />
    <span
      className="MuiSlider-track"
      style={
        Object {
          "left": "0%",
          "width": "0%",
        }
      }
    />
    <input
      type="hidden"
      value="0"
    />
    <span
      aria-labelledby="continuous-slider"
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={0}
      className="MuiSlider-thumb makeStyles-sliderThumb-3 MuiSlider-thumbColorPrimary"
      data-index={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      role="slider"
      style={
        Object {
          "left": "0%",
        }
      }
      tabIndex={0}
    />
  </span>
  <div
    className="makeStyles-sliderLimits-6"
  >
    <span
      className="MuiTypography-root makeStyles-disabledText-4 MuiTypography-overline"
    >
      0
    </span>
    <span
      className="MuiTypography-root makeStyles-disabledText-4 MuiTypography-overline"
      style={
        Object {
          "textAlign": "end",
        }
      }
    >
      100.00
    </span>
  </div>
  <p
    className="MuiFormHelperText-root Mui-required"
  />
</div>
`;
